client/src/rubin/nublado/client/_http.py

```diff
@@ -0,0 +1,557 @@
+        fetch_mode
+            Value of ``Sec-Fetch-Mode`` header to send. This suppresses some
+            log noise and improves XSRF handling in JupyterHub and JupyterLab.
+
+        Returns
+        -------
+        httpx.Response
+            HTTP response at the end of any redirect chain.
+
+        Raises
+        ------
+        NubladoDiscoveryError
+            Raised if Nublado is missing from service discovery.
+        NubladoRedirectError
+            Raised if the redirect leaves the Nublado URL space.
+        """
+        url = await self._url_for(route)
+        headers = await self._headers_for(route, fetch_mode=fetch_mode)
+        return await self._get(url, headers)
+
+    async def open_sse_stream(
+        self, route: str
+    ) -> AbstractAsyncContextManager[EventSource]:
+        """Open a server-sent events stream.
+
+        Parameters
+        ----------
+        route
+            Route relative to the base URL of Nublado. Routes starting with
+            ``user`` are considered JupyterLab routes.
+
+        Returns
+        -------
+        contextlib.AbstractAsyncContextManager
+            Context manager that provides an ``httpx_sse.EventSource``.
+
+        Raises
+        ------
+        NubladoDiscoveryError
+            Raised if Nublado is missing from service discovery.
+        """
+        url = await self._url_for(route)
+        headers = await self._headers_for(route, add_referer=True)
+        return aconnect_sse(self._client, "GET", url, headers=headers)
+
+    async def open_websocket(
+        self,
+        route: str,
+        *,
+        open_timeout: timedelta,
+        max_size: int | None,
+    ) -> AbstractAsyncContextManager[ClientConnection]:
+        """Open a WebSocket connection.
+
+        Parameters
+        ----------
+        route
+            Route relative to the base URL of Nublado. Routes starting with
+            ``user`` are considered JupyterLab routes.
+        max_size
+            Maximum size of a WebSocket message, or `None` for no limit.
+
+        Returns
+        -------
+        contextlib.AbstractAsyncContextManager
+            Context manager that provides a
+            `~websockets.asyncio.client.ClientConnection`.
+
+        Raises
+        ------
+        NubladoDiscoveryError
+            Raised if Nublado is missing from service discovery.
+        """
+        url = await self._url_for(route)
+        headers = await self._headers_for(route)
+        request = self._client.build_request("GET", url, headers=headers)
+        headers["Cookie"] = request.headers["Cookie"]
+        return websockets.connect(
+            self._url_for_websocket(url),
+            additional_headers=headers,
+            open_timeout=open_timeout.total_seconds(),
+            max_size=max_size,
+        )
+
+    async def post(
+        self,
+        route: str,
+        *,
+        content: str | None = None,
+        data: dict[str, Any] | None = None,
```

Comment:
Not that it matters, as it's a "stop doing that then" but what happens if you use both the data and the json arguments?  The httpx documentation doesn't directly state, at least not in the obvious page about how to use data and json.

Reply:
https://github.com/encode/httpx/blob/master/httpx/_content.py#L186

Data wins if both data and json are specified.